cupy/_manipulation/add_remove.py

```diff
@@ -169,15 +174,114 @@ def unique(ar, return_index=False, return_inverse=False,
             * The number of times each of the unique values comes up in the
               original array. Only provided if `return_counts` is True.
 
+    Notes:
+       When an axis is specified the subarrays indexed by the axis are sorted.
+       This is done by making the specified axis the first dimension of the
+       array (move the axis to the first dimension to keep the order of the
+       other axes) and then flattening the subarrays in C order.
+
     .. warning::
 
         This function may synchronize the device.
 
     .. seealso:: :func:`numpy.unique`
     """
-    if axis is not None:
-        raise NotImplementedError('axis option is not supported yet.')
+    if axis is None:
+        ret = _unique_1d(ar, return_index=return_index,
+                         return_inverse=return_inverse,
+                         return_counts=return_counts,
+                         equal_nan=equal_nan)
+        return ret
+
+    ar = cupy.moveaxis(ar, axis, 0)
+
+    # The array is reshaped into a contiguous 2D array
+    orig_shape = ar.shape
+    idx = cupy.arange(0, orig_shape[0], dtype=cupy.intp)
+    ar = ar.reshape(orig_shape[0], math.prod(orig_shape[1:]))
+    ar = cupy.ascontiguousarray(ar)
+
+    def compare_axis_elems(idx1, idx2):
+        for i in range(0, ar.shape[1]):
+            left = ar[idx1, i]
+            right = ar[idx2, i]
+
+            if cupy.isnan(left) and cupy.isnan(right):
```

Comment:
I'm now sure which should be the behaviour regarding NaNs when `unique` is called with a given axis

Reply:
Maybe it's worth checking the NumPy behavior? Note that NumPy has a new kw-only arg `equal_nan`.

Reply:
The NumPy behaviour is not consistent across datatypes, sometimes `NaN`s get pushed to the beginning on integral types and sometimes to the end on floating types. I'm not sure how to reconcile this two types of behaviours.

Reply:
This is the output of the comparison tests when `equal_nan` is passed:

![imagen](https://user-images.githubusercontent.com/1878982/187954433-e4e73d64-7aa4-4c85-a2e3-7a09cdc4cb08.png)

For instance, `float32` types have a consistent return when they are compared against NumPy, however, other types such as `float16` and `complex` types behave differently:

## float16 - equal_nan=True
```
cupy_result = array([[ 2., nan,  2.],
       [nan,  1., nan],
       [nan,  1., nan]], dtype=float16)

numpy_result = array([[nan,  1., nan],
       [nan,  1., nan],
       [ 2., nan,  2.]], dtype=float16)
```

As you can see, when `equal_nan=True`, then `NaNs` get pushed to the front of the result, whereas when `equal_nan=False`, they get pushed to the end. 

Given that this PR makes `NaNs` greater when `equal_nan=True`, that means that they will get pushed to the end rather to the beginning. However, as it is seen on the examples, NumPy sometimes pushes them to the front and others to the end, depending on the datatype

Reply:
@andfoy some testing shows that `numpy` has a bug for `float16`, and all other dtypes seem fine:
```
>>> xp = np
>>> a = xp.array([[2, xp.nan, 2],
...               [xp.nan, 1, xp.nan],
...               [xp.nan, 1, xp.nan]])
>>> np.unique(a.astype(np.float64), axis=0)
array([[ 2., nan,  2.],
       [nan,  1., nan],
       [nan,  1., nan]])
>>> np.unique(a.astype(np.float32), axis=0)
array([[ 2., nan,  2.],
       [nan,  1., nan],
       [nan,  1., nan]], dtype=float32)
>>> np.unique(a.astype(np.float128), axis=0)
array([[ 2., nan,  2.],
       [nan,  1., nan],
       [nan,  1., nan]], dtype=float128)
>>> np.unique(a.astype(np.float16), axis=0)
array([[nan,  1., nan],
       [nan,  1., nan],
       [ 2., nan,  2.]], dtype=float16)
>>> np.unique(a.astype(np.complex256), axis=0)
array([[ 2.+0.j, nan+0.j,  2.+0.j],
       [nan+0.j,  1.+0.j, nan+0.j],
       [nan+0.j,  1.+0.j, nan+0.j]], dtype=complex256)
>>> np.unique(a.astype(np.complex128), axis=0)
array([[ 2.+0.j, nan+0.j,  2.+0.j],
       [nan+0.j,  1.+0.j, nan+0.j],
       [nan+0.j,  1.+0.j, nan+0.j]])
>>> np.unique(a.astype(np.complex64), axis=0)
array([[ 2.+0.j, nan+0.j,  2.+0.j],
       [nan+0.j,  1.+0.j, nan+0.j],
       [nan+0.j,  1.+0.j, nan+0.j]], dtype=complex64)
```

CuPy does the right thing here, and NumPy has a bug specific to `float16`. That is true independent of the value of `equal_nan`. So I suggest to open a new issue on the NumPy issue tracker, and skip the `float16` tests in this PR.

The second issue is the failures for complex dtypes with `equal_nan=True`. I don't have a `cupy` install at hand, can you please post a simple reproducer like mine above?